providers/azure.go

```diff
@@ -0,0 +1,128 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/openpubkey/openpubkey/discover"
+)
+
+// AzureOptions is an options struct that configures how providers.AzureOp
+// operates. See providers.GetDefaultAzureOpOptions for the recommended default
+// values to use when interacting with Azure as the OpenIdProvider.
+type AzureOptions struct {
+	// ClientID is the client ID of the OIDC application. It should be the
+	// expected "aud" claim in received ID tokens from the OP.
+	ClientID string
+	// Issuer is the OP's issuer URI for performing OIDC authorization and
+	// discovery.
+	Issuer string
+	// Scopes is the list of scopes to send to the OP in the initial
+	// authorization request.
+	Scopes []string
+	// RedirectURIs is the list of authorized redirect URIs that can be
+	// redirected to by the OP after the user completes the authorization code
+	// flow exchange. Ensure that your OIDC application is configured to accept
+	// these URIs otherwise an error may occur.
+	RedirectURIs []string
+	// GQSign denotes if the received ID token should be upgraded to a GQ token
+	// using GQ signatures.
+	GQSign bool
+	// OpenBrowser denotes if the client's default browser should be opened
+	// automatically when performing the OIDC authorization flow. This value
+	// should typically be set to true, unless performing some headless
+	// automation (e.g. integration tests) where you don't want the browser to
+	// open.
+	OpenBrowser bool
+	// HttpClient is the http.Client to use when making queries to the OP (OIDC
+	// code exchange, refresh, verification of ID token, fetch of JWKS endpoint,
+	// etc.). If nil, then http.DefaultClient is used.
+	HttpClient *http.Client
+	// IssuedAtOffset configures the offset to add when validating the "iss" and
+	// "exp" claims of received ID tokens from the OP.
+	IssuedAtOffset time.Duration
+	// TenantID is the GUID  of the Azure tenant/organization. Azure has a
+	// different issuer URI for each tenant. Users that are not part of Azure
+	// organization, which microsoft nicknames consumers have a default
+	// tenant ID of "9188040d-6c67-4c5b-b112-36a304b66dad"
+	// More details can be found at
+	// https://learn.microsoft.com/en-us/entra/identity-platform/access-tokens
+	TenantID string
+}
+
+func GetDefaultAzureOpOptions() *AzureOptions {
+	defaultTenantID := "9188040d-6c67-4c5b-b112-36a304b66dad"
+	return &AzureOptions{
+		Issuer:   azureIssuer(defaultTenantID),
+		ClientID: "bd345b9c-6902-400d-9e18-45abdf0f698f", // TODO: replace with a better client ID
```

Comment:
I wonder if we should really have a default client ID :thinking: Perhaps this should be required to be set by the calling code, WDYT? (Probably not for this PR as the google one also has a hard-coded default one.)

Reply:
I want someone to just be able to download OpenPubkey and use it out of the box. Without a default client ID they would first need to sign into Azure or GCP and go through that whole process and then configure the clientID in the provider.

With OPKSSH we should be able to be able to use a configure file for this.